mmengine/model/wrappers/pipeline_distributed.py

```diff
@@ -0,0 +1,1199 @@
+                 num_pipelines: Optional[int] = None,
+                 num_chunks: Optional[int] = None,
+                 memory_threshold: float = 0.7,
+                 memory_map: Optional[Dict[str, str]] = None,
+                 no_split_module_classes: Optional[List[str]] = None,
+                 language_module_class: Optional[str] = None,
+                 device_map: Union[str, Dict[str, dict]] = 'auto',
+                 offload_directory: Optional[str] = None,
+                 exec_entry: str = '__call__'):
+
+        if digit_version(TORCH_VERSION) <= digit_version('2.0.0'):
+            raise RuntimeError(
+                'MMPipelineParallel should work with PyTorch >= 2.0.0')
+
+        super().__init__()
+
+        # init model
+        self.weights = weights
+        self.model_state_dict: Dict[str, torch.Tensor] = {}
+        self.model = self._init_model(model)
+        self.entry = getattr(self.model, exec_entry)
+        # init pipeline parallelism
+        if num_pipelines is not None:
+            self.num_pipelines = num_pipelines
+            if self.num_pipelines > torch.cuda.device_count():
+                warnings.warn('The number of pipelines is larger than ' +
+                              'the number of GPUs. ' +
+                              'There may be some unpredictable bugs.')
+        else:
+            self.num_pipelines = torch.cuda.device_count()
+        if num_chunks is not None:
+            self.num_chunks = num_chunks
+        else:
+            self.num_chunks = self.num_pipelines * 32
+
+        # init memory map
+        self.memory_threshold = memory_threshold
+        self.memory_map = self._init_memory_map(memory_map)
+
+        # init device map
+        self.no_split_module_classes = [] if no_split_module_classes is None \
+            else no_split_module_classes
+        self.language_module_class = language_module_class
+        self.lm_offset = 0
+        self.model_tree = self._get_model_tree()
+        if isinstance(device_map, dict):
+            self.device_map = device_map
+            self.offload_map = self._init_offload_map()
+            self.module_map = self._init_module_map()
+            self._load_and_dispatch(self.weights)
+            self._register_hooks()
+            MMPipelineParallel.stream_contexts = self._init_stream_contexts()
+            self._inited = True
+        else:
+            # after we get the input shape, we can init the device map
+            self.device_map_policy = device_map
+            self._inited = False
+
+        # init offload directory
+        if offload_directory is not None:
+            self.offload_directory = offload_directory
+        else:
+            self.offload_directory = tempfile.mkdtemp()
+        self.offloaded_weights: Dict[str, Dict[str, Any]] = {}
+
+        # init queues
+        MMPipelineParallel.in_queues, MMPipelineParallel.out_queues = \
+            self._init_queues()
```

Comment:
Why do we not assign them to self?

```python
self.in_queues, self.out_queues =self._init_queues()
```

Reply:
Since these variables may be called by other in-class functions, such as `MMPipelineParallelism.Hook`, calling `self.events` at `MMPipelineParallelism.Hook._exit_prev_part` is wrong. Therefore, in order to standardize the writing style, the in-class class (`Hook` and `Flag`), these variables and static functions should be called by `MMPipelineParallel.xxxx`.